### ISSUE TITLE
OSError: [Errno 5] Input/output error: 'lock.stats.tsv'

### ISSUE THREAD

**Issue by dearxxj** [databio/pepatac#116]
Hi, just curious that anyone had this error? I followed the tutorial(#122-install-software-requirements-natively) to install the pipeline and I think I got everything right. 

Below is the error message. Also, the log file is attached.

> `File_mb`	4461.7806	PEPATAC	_RES_
Traceback (most recent call last):
  File "/hpc/home/abc/src/pepatac/pipelines/pepatac.py", line 1422, in <module>
    sys.exit(main())
  File "/hpc/home/abc/src/pepatac/pipelines/pepatac.py", line 543, in main
    [x for x in [args.input, args.input2] if x is not None]))
  File "/hpc/home/abc/python/lib/python2.7/site-packages/pypiper/manager.py", line 1243, in report_result
    self._safe_write_to_file(self.pipeline_stats_file, message_raw)
  File "/hpc/home/abc/python/lib/python2.7/site-packages/pypiper/manager.py", line 1322, in _safe_write_to_file
    os.remove(lock_file)
OSError: [Errno 5] Input/output error: 'ATAC-seq/pepatac/868/lock.stats.tsv'
Pipeline status: running
Starting cleanup: 0 files; 0 conditional files for cleanup

I think it may be a pypiper error. Any comments would be greatly appreciated!

[PEPATAC_log.txt](https://github.com/databio/pepatac/files/3297723/PEPATAC_log.txt)

**Comment by nsheff** [comment]
That's a strange error... if you run it again  after deleting the output folder completely, do you get the same error?

Can you confirm that you have permissions set correctly on the 2 input files?

**Comment by dearxxj** [comment]
> That's a strange error... if you run it again after deleting the output folder completely, do you get the same error?
> 
> Can you confirm that you have permissions set correctly on the 2 input files?

Thanks for the quick reply! I tried deleting the output folder completely and I still got the same error. I have the read permission to the 2 input files. The permission is set as "-rw-rw-r--".

**Comment by nsheff** [comment]
this is quite strange. I can't reproduce it using local files here...the error is coming from this line: 

https://github.com/databio/pepatac/blob/a0e4347b199c91bbfd7d994b0705da2ca8d51015/pipelines/pepatac.py#L543

ultimately to this line:

https://github.com/databio/pypiper/blob/master/pypiper/manager.py#L1322

But I can't find any information on why `os.remove` would be throwing this kind of error.  It looks like it should be correct to me...

Do you also get this error with different file inputs? What OS are you using?

**Comment by nsheff** [comment]
can you also say whether the lock file exists and what it's permissions are ('ATAC-seq/pepatac/868/lock.stats.tsv'), 

and also if the "File_mb" line was correctly printed in the `stats.tsv` file?

**Comment by nsheff** [comment]
here's an interesting thought: https://stackoverflow.com/a/43518633/946721

>  I had this problem at university when I exceeded my quota. You can check your quota with `quota -s`

**Comment by dearxxj** [comment]
I tried different datasets and error is the same. My OS is 

> Linux hpcgateway01 2.6.32-754.12.1.el6.x86_64 #1 SMP Thu Mar 7 22:07:44 EST 2019 x86_64 x86_64 x86_64 GNU/Linux

The `stats.tsv` file has only one line:

> File_mb 4461.7806       PEPATAC

`quota -s` gives me an error, but I'm pretty sure my quota has not run out because I can write other files.

**Comment by nsheff** [comment]
Ok, thanks. I think this has something to do with your local HPC setup. Can you ask the people in charge there and let me know what they say? 

I can't find anything wrong with the software and it works fine in multiple other computing environments...

you might be able to narrow it down to just starting up python and typing:

```
import os
os.remove("lock.stats.tsv")
```

Why can't you remove that file?

**Comment by dearxxj** [comment]
It's strange that If I just run os.remove(), the file can be removed without an error...

`python -c 'import os;os.remove("lock.stats.tsv")'`

I guess that file is used by other programs while being removed by pypiper.

**Comment by nsheff** [comment]
Hmm... that is strange. Try something like this:

```
import pypiper
pm = pypiper.PipelineManager("test", "testfolder")
pm._safe_write_to_file("test.txt", "message")
pm.stop_pipeline()
```

**Comment by dearxxj** [comment]
> Hmm... that is strange. Try something like this:
> 
> ```
> import pypiper
> pm = pypiper.PipelineManager("test", "testfolder")
> pm._safe_write_to_file("test.txt", "message")
> pm.stop_pipeline()
> ```

This line `pm._safe_write_to_file("test.txt", "message")` replicates the error. I tried to uninstall and re-install pypiper and it didn't work.

**Comment by nsheff** [comment]
Ok, tell me if this replicates the error:

```
import os
file = "test.txt"
write_lock_flags = os.O_CREAT | os.O_EXCL | os.O_WRONLY
f = os.open(file, write_lock_flags)
os.remove(file)
```

and if this solves it:

```
import os
file = "test.txt"
write_lock_flags = os.O_CREAT | os.O_EXCL | os.O_WRONLY
f = os.open(file, write_lock_flags)
os.close(f)
os.remove(file)
```

**Comment by nsheff** [comment]
if that works, you can try updating to the dev pypiper with: 

`pip install --user --upgrade https://github.com/databio/pypiper/archive/dev.zip`

and let me know if that works.

**Comment by dearxxj** [comment]
> if that works, you can try updating to the dev pypiper with:
> 
> `pip install --user --upgrade https://github.com/databio/pypiper/archive/dev.zip`
> 
> and let me know if that works.

Yes! This solves the problem perfectly! Now the pipeline is running without any error. Thank you so much for your help.

**Comment by nsheff** [comment]
great. This will be fixed then in the next release version of pypiper (0.11.3). Thanks for reporting it!

**Comment by nsheff** [comment]
FYI, it's now in the release version of pypiper.

**Comment by dearxxj** [comment]
> FYI, it's now in the release version of pypiper.

That's great! I'll try it. Thanks!

**Comment by dearxxj** [comment]
As the pipeline was running, there is another error that popped out. I was able to fix it by myself, though not sure if I did it right. I feel that it's not necessary to open a new issue, and I'll just paste it here.

> 'pigz -f -p 5 -d -c pepatac/868/raw/hg38_annotations.bed.gz | cut -f 4 | sort -u'
Traceback (most recent call last):
  File "/hpc/home/abc/src/pepatac/pipelines/pepatac.py", line 1422, in <module>
    sys.exit(main())
  File "/hpc/home/abc/src/pepatac/pipelines/pepatac.py", line 1280, in main
    ftList = str.splitlines(ftList)
TypeError: descriptor 'splitlines' requires a 'str' object but received a 'unicode'

The error occurs at line 1280 of the pepatac.py file.
https://github.com/databio/pepatac/blob/a0e4347b199c91bbfd7d994b0705da2ca8d51015/pipelines/pepatac.py#L1280

I think it's because I'm using python2 and str is not a unicode string in python2 (found [here](https://stackoverflow.com/questions/21808657/what-is-a-unicode-string)). I was able to fix it by changing it to
`ftList = str.splitlines(str(ftList))`
